File: NQueens.java
import java.util.*;

class Solution {

    boolean isSafe(int row, int col, List<String> board) {

        int r = row;
        int c = col;

        // Upper
        while (r >= 0) {

            if (board.get(r).charAt(c) == 'Q') {
                return false;
            }

            r--;
        }

        r = row;
        c = col;

        // Upper-left diagonal
        while (r >= 0 && c >= 0) {

            if (board.get(r).charAt(c) == 'Q') {
                return false;
            }

            r--;
            c--;
        }

        r = row;
        c = col;

        // Upper-right diagonal
        while (r >= 0 && c < board.size()) {

            if (board.get(r).charAt(c) == 'Q') {
                return false;
            }

            r--;
            c++;
        }

        return true;
    }

    void recursion(int row,
                   List<List<String>> ans,
                   List<String> board,
                   int n) {

        if (row == board.size()) {

            ans.add(new ArrayList<>(board));
            return;
        }

        for (int col = 0; col < n; col++) {

            if (isSafe(row, col, board)) {

                char[] currentRow = board.get(row).toCharArray();
                currentRow[col] = 'Q';
                board.set(row, new String(currentRow));

                recursion(row + 1, ans, board, n);

                currentRow = board.get(row).toCharArray();
                currentRow[col] = '.';
                board.set(row, new String(currentRow));
            }
        }
    }

    public List<List<String>> NQueens(int n) {

        List<String> board = new ArrayList<>();

        for (int i = 0; i < n; i++) {

            String row = ".".repeat(n);
            board.add(row);
        }

        List<List<String>> ans = new ArrayList<>();

        recursion(0, ans, board, n);

        return ans;
    }
}

public class Main {

    public static void main(String[] args) {

        Solution solution = new Solution();

        int n = 4;

        List<List<String>> solutions = solution.NQueens(n);

        for (List<String> sol : solutions) {

            for (String row : sol) {
                System.out.println(row);
            }

            System.out.println();
        }
    }
}

File: NQueens.cpp
// Online C++ compiler to run C++ program online
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution{
    
    bool isSafe(int row, int col, vector<string>&board){
        int r = row;
        int c = col;
        
        //upper
        while(r >= 0){
            if(board[r][c] == 'Q') return false;
            r--;
        }
        
         r = row;
         c = col;
        
        //upper-left
        while(r >= 0 && c >= 0){
            if(board[r][c] == 'Q') return false;
            r--;
            c--;
        }
        
         r = row;
         c = col;
        
        while(r >= 0 && c < board.size()){
            if(board[r][c] == 'Q') return false;
            r--;
            c++;
        }
        
        return true;
    }
    
    void recursion(int row, vector<vector<string>>&ans, vector<string>&board, int n){
        
        if(row == board.size()){
            ans.push_back(board);
            return;
        }
        
        for(int col = 0; col < n; col++){
            if(isSafe(row, col, board)){
                board[row][col] = 'Q';
                recursion(row+1, ans, board, n);
                board[row][col] = '.';
            }
        }
    }
    
    public:
    vector<vector<string>> NQueens(int n){
        vector<string>board(n, string(n, '.'));
        vector<vector<string>> ans;
        
        recursion(0, ans, board, n);
        
        return ans;
        
        
    }
};

int main() {
    
    Solution solution;
    int n = 4; 
    vector<vector<string>> solutions = solution.NQueens(n);

    
    for (const auto& sol : solutions) {
        for (const auto& row : sol) {
            cout << row << endl;
        }
        cout << endl;
    }

    
    return 0;
}
